android/greengrass-android-component-base/component-library/src/main/java/com/aws/greengrass/android/component/core/GreengrassComponentService.java

```diff
@@ -0,0 +1,876 @@
+    public void onCreate() {
+        try {
+            Log.d(LOG_TAG, "onCreate");
+            super.onCreate();
+            messengerThread.get().start();
+            registerBroadcastReceiver();
+        } catch (Throwable e) {
+            Log.e(LOG_TAG, "Exception " + Log.getStackTraceString(e));
+        }
+    }
+
+    @Override
+    public void onDestroy() {
+        try {
+            Log.d(LOG_TAG, "onDestroy");
+            unregisterReceiver(receiver);
+            tryStop(EXIT_CODE_KILLED);
+            MessengerLooperThread msgThread = messengerThread.getAndSet(null);
+            msgThread.finish();
+        } catch (Throwable e) {
+            Log.e(LOG_TAG, "Exception " + Log.getStackTraceString(e));
+        } finally {
+            // In any case trigger the callback and destroy base object
+            onServiceStopped();
+            super.onDestroy();
+        }
+    }
+
+    /**
+     * Greengrass Service Stopped event callback
+     * This event callback is triggered after Greengrass component is fully stopped and all the
+     * communication with Nucleus is done, including the report of exit code. You can use this
+     * callback to do some final cleanup.
+     * Default implementation terminates all the IPC connections. Override this method if your
+     * application contains multiple Greengrass services and you need to keep IPC working after one
+     * of the services exits
+     */
+    protected void onServiceStopped() {
+        // By default close all existing IPC connections.
+        IPCUtils.closeAllConnections();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public int onStartCommand(Intent intent, int flags, int startId) {
+        boolean startupLockAcquired = false;
+        try {
+            // Acquire lock to avoid race condition between startService() and bind()
+            Log.d(LOG_TAG, "onStartCommand - trying to acquire lock");
+            startupLockAcquired = serviceStartupLock.tryAcquire(STARTUP_LOCK_WAIT_LIMIT_MS, TimeUnit.MILLISECONDS);
+            if (!startupLockAcquired) {
+                throw new TimeoutException("onStartCommand - unable to acquire startup lock within timeout");
+            }
+            // Proceed
+            Log.d(LOG_TAG, "onStartCommand - lock acquired");
+            if (intent == null) {
+                // TODO: handle restart here
+            } else if (ACTION_START_COMPONENT.equals(intent.getAction())) {
+                Log.d(LOG_TAG, "got START command");
+                startInForeground();
+                if (ensureStopped()) {
+                    masterPackage = intent.getStringExtra(EXTRA_MASTER_PACKAGE);
+                    arguments = intent.getStringArrayExtra(EXTRA_ARGUMENTS);
+
+                    // Try to extract Environment supplied by Nucleus
+                    if (intent.hasExtra(EXTRA_COMPONENT_ENVIRONMENT)) {
+                        HashMap<String, String> env = (HashMap<String, String>) intent
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Use `Parcelable`, not `Serializable`, to pass data between different components in Android. `Parcelable` is more performant. See the following diff on GitHub that makes this code change: [diff](https://github.com/k9mail/k-9/commit/685df0ff6772df853457e782809ab3d064f533d5#diff-3e4082b13ee86b8238eff49b864ff82db653e5af02ce4ee9c7e21b6c9bfca3a2L330)

Learn more: [Parcelable in the Android API reference](https://developer.android.com/reference/android/os/Parcelable) and [Serializable and Parcelable in Android](https://skoric.svbtle.com/serializable-vs-parcelable)